### PR TITLE
Support for lazily loading multimedia in Commcare

### PR DESCRIPTION
**Pull request by AmitPhulera**
https://dimagi-dev.atlassian.net/projects/ICDS/issues/ICDS-1402
##### SUMMARY
Added attribute `lazy` on Media resources in the `media_suite.xml`
##### FEATURE FLAG
LAZY_LOAD_MULTIMEDIA

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml [hunk 1]
```diff
@@ -8,6 +8,15 @@
   since: "1.3"
   force: true
 
+- name: "Lazy Load Resources"
+  description: "How should Commcare load the the multimedia files in the application."
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:12]: It isn't clear what the default is. Can you rephrase to something like "Load multimedia files lazily"?

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml [hunk 2]
```diff
@@ -8,6 +8,15 @@
...
+  id: "lazy_load_multimedia_files"
+  values: ["true", "false"]
+  value_names: ["Yes", "No"]
+  default: "false"
+  since: "2.5"
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:17]: Why is this 2.5? Wasn't this added recently, maybe 2.49?

**Review by orangejenny — COMMENTED** [review]
Looks reasonable overall, left a few minor comments.

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml (hunk 2 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:17]: Think it will be released in 2.5, I confirmed it from @shubham1g5

Comment (shubham1g5) [reply corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:17]: it should be 2.50

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml (hunk 1 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml]: Done

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml (hunk 2 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml]: Version updated to 2.50

File: corehq/apps/app_manager/static_strings.py [hunk 3]
```diff
@@ -135,6 +137,7 @@
     ugettext_noop('Must correspond to the password format specified below.'),
     ugettext_noop('Native (International)'),
     ugettext_noop('Never'),
+    ugettext_noop('No'),
```
Comment (orangejenny) [inline corehq/apps/app_manager/static_strings.py:140]: You don't need this; it's a few lines down.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/app_manager/static_strings.py (hunk 3 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/static_strings.py]: I missed it.
Thanks for pointing out 😄

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/app_manager/suite_xml/generator.py [hunk 4]
```diff
@@ -148,6 +148,10 @@ def media_resources(self):
             install_path = '../../{}'.format(path)
             local_path = './{}/{}'.format(path, name)
 
+            load_lazily = False
+            lazy_load_preference = self.app.profile.get('properties', {}).get('lazy-load-multimedia-files')
+            if lazy_load_preference and lazy_load_preference == 'true':
+                load_lazily = True
```
Comment (mkangia) [inline corehq/apps/app_manager/suite_xml/generator.py:154]: You could replace this with
`load_lazily = self.app.profile.get('properties', {}).get('lazy-load-multimedia-files') == 'true'`
or
```
load_lazily = (
    self.app.profile.get('properties', {}).
    get('lazy-load-multimedia-files') == 'true'
)
```
or if you like the break down
```
lazy_load_preference = self.app.profile.get('properties', {}).get('lazy-load-multimedia-files')
load_lazily = lazy_load_preference == 'true'
```

**Review by mkangia — APPROVED** [review]

**Comment by shubham1g5** [comment]
Taking the conversation on whether to implement a Feature flag vs custom settings here. In my opinion the custom setting `Load multimedia files lazily` should be behind a feature flag so that a normal non enterprise project is not able to accidently turn on the setting and face issues. I also think this feature is mostly relevant to CAS given the bandwidth issues being a bottleneck for app updates. Other projects might misunderstood the expectation that the app can be presented to the user without any media at all.  Also curious to understand if there are any HQ standards on when to implement something as a feature flag vs custom setting ?

**Review by orangejenny — CHANGES_REQUESTED** [review]
Marking as request changes because, as discussed on spec, I completely forgot that this should be only for videos.

**Comment by orangejenny** [comment]
> Also curious to understand if there are any HQ standards on when to implement something as a feature flag vs custom setting ?

After offlining with @shubham1g5 I've come around to the belief that this should be a feature flag. There isn't a clear standard for flags versus app settings, though flags tend to be for feature that are harder to understand and/or less likely to be useful to many projects.

This feature as implemented I think is not hard to understand and may well be useful to other projects. However, since I now understand we're likely to add functionality in the future to specify laziness at the file level, I'd prefer it go behind a flag.

With file-level laziness, this feature could still be useful to other projects, but that'll be a bigger UI change, so it shouldn't go out to all users unless product has had a chance to understand it and agree the approach makes sense for our whole user base. I don't want that future work to hinge on how available the product team is, so it's safer to just flag it now.

File: corehq/apps/app_manager/suite_xml/generator.py (hunk 4 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/suite_xml/generator.py]: The second snippet looks real nice, implemented that. 
Thanks for the suggestion. 😄

**Comment by AmitPhulera** [comment]
Hey @orangejenny, 
I have updated the PR to add support for lazy loading for video only.

**Comment by AmitPhulera** [comment]
Hey @orangejenny, 
@shubham1g5 confirmed that things look good on staging, re-requesting your review on it just to be sure everything looks good from your end.

File: corehq/toggles.py [hunk 5]
```diff
@@ -439,6 +439,13 @@ def _ensure_valid_randomness(randomness):
     [NAMESPACE_DOMAIN],
 )
 
+OPTIONAL_MEDIA_IN_UPDATES = StaticToggle(
```
Comment (mkangia) [inline corehq/toggles.py:442]: the name of this and the feature are different, I see people searching for "lazy video" and then not finding it. You might want to reconsider this name

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml [hunk 6]
```diff
@@ -8,9 +8,10 @@
   since: "1.3"
   force: true
 
-- name: "Load multimedia files lazily"
-  description: "How should Commcare load the the multimedia files in the application. By default Commcare will try to load them instantly"
-  id: "lazy_load_multimedia_files"
+- name: "Load video files lazily"
+  description: "How should Commcare load the the video files in the application. By default Commcare will try to load them instantly"
+  id: "lazy_load_video_files"
+  toggle: OPTIONAL_MEDIA_IN_UPDATES
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:14]: I didn't know you could put toggle here? I don't see toggle elsewhere.

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml [hunk 7]
```diff
@@ -8,6 +8,16 @@
...
 
+- name: "Load video files lazily"
+  description: "Make OTA app updates skip download for video files at app update time."
+  id: "lazy_load_video_files"
+  toggle: OPTIONAL_MEDIA_IN_UPDATES
+  values: ["true", "false"]
+  value_names: ["Yes", "No"]
+  default: "false"
+  since: "2.50"
+  force: true
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:19]: just curious, what does `force: true` does?

**Review by mkangia — APPROVED** [review]
Couple of minor comments which i think should be addressed. Will wait for @orangejenny to give a final go ahead.

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml (hunk 6 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:14]: Yeah, this works: https://github.com/dimagi/commcare-hq/blob/1c770f3c1e9487470349bda3ded952f5fdbcdbb4/corehq/apps/app_manager/views/apps.py#L185-L187

There are good [docs](urlhttps://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/commcaresettings-README.md) for commcare settings options.

File: corehq/apps/app_manager/suite_xml/generator.py [hunk 8]
```diff
@@ -134,7 +134,7 @@ def generate_suite(self):
     def media_resources(self):
         PREFIX = 'jr://file/'
         multimedia_map = self.app.multimedia_map_for_build(build_profile=self.build_profile, remove_unused=True)
-        lazy_load_preference = self.app.profile.get('properties', {}).get('lazy-load-video-files')
```
Comment (orangejenny) [inline corehq/apps/app_manager/suite_xml/generator.py]: Did @shubham1g5 ask for this change? I thought property names were usually dashed case, but he would know for sure.

**Review by orangejenny — APPROVED** [review]
Looks good to me. Nice well-factored commits.

File: corehq/apps/app_manager/suite_xml/generator.py (hunk 8 above)
Comment (shubham1g5) [reply corehq/apps/app_manager/suite_xml/generator.py]: Agree that it should be dashed-case to maintain consistency.

Comment (mkangia) [reply corehq/apps/app_manager/suite_xml/generator.py]: Github shows that comment on the old code guess because it was commented on that.
Just in case someone gets confused too, it was changed to `lazy_load_video_files` and hence the comment.

File: corehq/toggles.py (hunk 5 above)
Comment (AmitPhulera) [reply corehq/toggles.py:442]: I used this name from the spec file. 
But you are right, it would be confusing for uses.
I would update it to `LAZY_LOAD_VIDEOS` and will update in the specs

File: corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml (hunk 6 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/static/app_manager/json/commcare-profile-settings.yaml:14]: Yep. I luckily stumbled upon the docs in code 😅

File: corehq/apps/app_manager/suite_xml/generator.py (hunk 8 above)
Comment (AmitPhulera) [reply corehq/apps/app_manager/suite_xml/generator.py]: So should it be dashed case?

File: corehq/toggles.py (hunk 5 above)
Comment (shubham1g5) [reply corehq/toggles.py:442]: @mkangia  @AmitPhulera The property `LAZY_LOAD_VIDEOS` is just one setting under the feature that makes all videos in an application optional, though the feature flag is aimed to provide cover for all such settings we roll out in future. If we have to make a change here, I think changing the property name to `mark-all-videos-optional` is a better change here instead of changing the feature flag.

File: corehq/apps/app_manager/suite_xml/generator.py (hunk 8 above)
Comment (shubham1g5) [reply corehq/apps/app_manager/suite_xml/generator.py]: right.

File: corehq/toggles.py (hunk 5 above)
Comment (AmitPhulera) [reply corehq/toggles.py:442]: Okay. Going with `mark-all-videos-optional`

Comment (AmitPhulera) [reply corehq/toggles.py:442]: @shubham1g5 Can we go with `LAZY_LOAD_MULTIMEDIA` feature flag, I belive it can encompass any other change that we do in the future. 

The setting name `mark-all-videos-optional` gives me a sense that we are making videos optional which we are not. They will eventually get downloaded to the app.

What are your views on it @mkangia @orangejenny ?

Comment (shubham1g5) [reply corehq/toggles.py:442]: no objections to `LAZY_LOAD_MULTIMEDIA` from my side.

**Review by mkangia — APPROVED** [review]
